.github/workflows/wheels.yml

```diff
@@ -36,235 +43,145 @@ on:
+        arch: ["x86_64", "universal2", "arm64"]
+        exclude:
+
+          # cp37 doesn't allow a univeral2 build
+          - py: cp37
+            arch: "universal2"
+
+          # arm64 isn't supported on macos github workers
+          - arch: "arm64"
 
     steps:
-      - name: Checkout ConfigSpace
+      - name: Checkout ${{ env.package-name }}
         uses: actions/checkout@v2
 
-      - name: Setup Python
-        uses: actions/setup-python@v2
+      - name: Build wheels with cibuildwheel to wheelhouse/*.whl
+        uses: pypa/cibuildwheel@v2.3.1
+        env:
+          CIBW_BUILD: ${{ matrix.py }}-*
+          CIBW_ARCHS: ${{ matrix.arch }}
+          CIBW_TEST_REQUIRES: ${{ env.test-reqs }}
+          CIBW_TEST_COMMAND: ${{ env.test-cmd }} {project}/${{ env.test-dir }}
 
-      - name: Build source distribution
-        run: python setup.py sdist
+      - name: Upload artifacts
+        uses: actions/upload-artifact@v2
+        with:
+          path: wheelhouse/*.whl
 
-        # Full unit-testing is handled in pytest.yml
-      - name: Twine check
-        run: |
-          pip install twine
-          last_dist=$(ls -t dist/ConfigSpace-*.tar.gz | head -n 1)
-          twine_output=`twine check "$last_dist"`
-          if [[ "$twine_output" != "Checking $last_dist: PASSED" ]]; then echo $twine_output && exit 1;fi
+  build_windows_wheels:
+    name: ${{ matrix.py }}-windows-${{ matrix.arch }}
+    runs-on: windows-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        py: [cp37, cp38, cp39, cp310]
+        arch: ["AMD64", "x86"]
+        exclude:
 
-      - name: Install dist
-        run: |
-          last_dist=$(ls -t dist/ConfigSpace-*.tar.gz | head -n 1)
-          pip install $last_dist
+          # We can't build win32 (x86) with cp310 because numpy doesn't have a win32 wheel
+          - py: cp310
+            arch: "x86"
 
-      - name: PEP 561 Compliance
-        run: |
-          pip install mypy
-          cd ..  # required to use the installed version of ConfigSpace
-          if ! python -c "import ConfigSpace"; then exit 1; fi
 
-      - name: Store artifacts
+    steps:
+      - name: Checkout ${{ env.package-name }}
+        uses: actions/checkout@v2
+
+      - name: Build wheels with cibuildwheel to wheelhouse/*.whl
+        uses: pypa/cibuildwheel@v2.3.1
+        env:
+          CIBW_BUILD: ${{ matrix.py }}-*
+          CIBW_ARCHS: ${{ matrix.arch }}
+          CIBW_TEST_REQUIRES: ${{ env.test-reqs }}
+          CIBW_TEST_COMMAND: ${{ env.test-reqs }} {project}/${{ env.test-dir }}
+
+      - name: Upload artifacts
         uses: actions/upload-artifact@v2
         with:
-          path: dist/*.tar.gz
+          path: wheelhouse/*.whl
 
-  # Upload the wheels and the source distribution
+  # Upload the wheels
   release_assets:
     name: Upload Release
     runs-on: ubuntu-latest
-    needs: [build_wheels, build_sdist]
+    needs: [build_linux_wheels, build_macos_wheels, build_windows_wheels]
```

Comment:
Hmm, if the wheels build and test fine then the src distribution should also be fine. The only difference is the twine check (which just checks that the README.md can be parsed by PyPi).

However, this is just my assumption and may as well just have it in code, I'll do this.

Reply:
Thanks a lot. Just checking, can we now have an upload from `dist.yaml`, but not `wheels.yaml` in case `dist.yaml` succeeds and `wheels.yaml` fails?

Reply:
I'm going to check on test-pypi if they can be both uploaded speratly or if they must be uploaded all together.

Reply:
It seems not, I think the best solution is requireing them all to pass to upload to PyPi

Reply:
> Note on the wheels and sdist, it seems they all have to be uploaded together. For that reason I think it's best to have them all in one workflow and uploaded together there. I don't see a use case where we only want to release the sdist for a version

As such, I think it makes the most sense to just rename files wheels.yml -> release.yml with the sdist checking in there and just remove sdist.yml completely. It's tested regularly in pytest.yml and the release version of it is tested in release.yml

release.yml also runs tests on both the wheels and sdist too

Lastly, if there ever is a use case to only upload the sdist and not wheels, this can just be manually downloaded and released, preventing any acceidental issues